From d98e0d7026c367d025b3f8114eacc7bf8ca29e06 Mon Sep 17 00:00:00 2001
From: markpavlitski <markpavlitski@gmail.com>
Date: Thu, 6 Jun 2013 16:29:54 +0100
Subject: [PATCH] Issue #2013785 by markpavlitski: Check if alias is a
 directory before deslashing.

---
 globalredirect.module | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/globalredirect.module b/globalredirect.module
index 8f24754..3e38a45 100644
--- a/globalredirect.module
+++ b/globalredirect.module
@@ -215,8 +215,10 @@ function globalredirect_init() {
   // Compare the request to the alias. This also works as a 'deslashing'
   // agent. If we have a language prefix then prefix the alias
   if ($request_path != $prefix . $alias) {
-    // If it's not just a slash or user has deslash on, redirect
-    if (str_replace($prefix . $alias, '', $request_path) != '/' || $settings['deslash']) {
+    // If it's not just a slash or user has deslash on, redirect, but avoid
+    // redirecting to a directory if we are only deslashing, as apache mod_dir
+    // can cause an infinite loop.
+    if (str_replace($prefix . $alias, '', $request_path) != '/' || ($settings['deslash'] && !is_dir($alias))) {
       globalredirect_goto($alias, $options);
     }
   }
-- 
1.8.1.2

