diff --git a/context/plugins/defaultcontent_condition.inc b/context/plugins/defaultcontent_condition.inc
new file mode 100644
index 0000000..76422da
--- /dev/null
+++ b/context/plugins/defaultcontent_condition.inc
@@ -0,0 +1,57 @@
+<?php
+
+class defaultcontent_condition extends context_condition {
+  /**
+   * Omit condition values. We will provide a custom input form for our conditions.
+   */
+  function condition_values() {
+    return array();
+  }
+
+  /**
+   * Condition form.
+   */
+  function condition_form($context) {
+    $form = parent::condition_form($context);
+    unset($form['#options']);
+
+    $form['#type'] = 'textarea';
+    $form['#default_value'] = implode("\n", $this->fetch_from_context($context, 'values'));
+    return $form;
+  }
+
+  /**
+   * Condition form submit handler.
+   */
+  function condition_form_submit($values) {
+    $parsed = array();
+    $items = explode("\n", $values);
+    if (!empty($items)) {
+      foreach ($items as $v) {
+        $v = trim($v);
+        if (!empty($v)) {
+          $parsed[$v] = $v;
+        }
+      }
+    }
+    return $parsed;
+  }
+
+  /**
+   * Execute.
+   */
+  function execute($node) {
+    if (!isset($node->machine_name)) {
+      return;
+    }
+    if ($this->condition_used()) {
+      foreach ($this->get_contexts() as $context) {
+        $machine_names = $this->fetch_from_context($context, 'values');
+
+        if (in_array($node->machine_name, $machine_names)) {
+          $this->condition_met($context);
+        }
+      }
+    }
+  }
+}
diff --git a/defaultcontent.module b/defaultcontent.module
index 0a94571..ba68d95 100644
--- a/defaultcontent.module
+++ b/defaultcontent.module
@@ -465,7 +465,7 @@ function defaultcontent_context_plugins() {
   $plugins = array();
     $plugins['defaultcontent_condition'] = array(
       'handler' => array(
-      'path' => drupal_get_path('module', 'defaultcontent') .'/plugins',
+      'path' => drupal_get_path('module', 'defaultcontent') .'/context/plugins',
       'file' => 'defaultcontent_condition.inc',
       'class' => 'defaultcontent_condition',
       'parent' => 'context_condition',
@@ -493,7 +493,7 @@ function defaultcontent_context_registry() {
  * Implements hook_node_view().
  */
 function defaultcontent_node_view($node, $view_mode, $langcode) {
-  if ($plugin = context_get_plugin('condition', 'defaultcontent')) {
+  if (module_exists('context') && $plugin = context_get_plugin('condition', 'defaultcontent')) {
     $plugin->execute($node);
   } 
 }
diff --git a/plugins/defaultcontent_condition.inc b/plugins/defaultcontent_condition.inc
deleted file mode 100644
index 10407cb..0000000
--- a/plugins/defaultcontent_condition.inc
+++ /dev/null
@@ -1,57 +0,0 @@
-<?php
-
-class defaultcontent_condition extends context_condition {
-  /**
-   * Omit condition values. We will provide a custom input form for our conditions.
-   */
-  function condition_values() {
-    return array();
-  }
-
-  /**
-   * Condition form.
-   */
-  function condition_form($context) {
-    $form = parent::condition_form($context);
-    unset($form['#options']);
-
-    $form['#type'] = 'textarea';
-    $form['#default_value'] = implode("\n", $this->fetch_from_context($context, 'values'));
-    return $form;
-  }
-  
-  /**
-   * Condition form submit handler.
-   */
-  function condition_form_submit($values) {
-    $parsed = array();
-    $items = explode("\n", $values);
-    if (!empty($items)) {
-      foreach ($items as $v) {
-        $v = trim($v);
-        if (!empty($v)) {
-          $parsed[$v] = $v;
-        }
-      }
-    }
-    return $parsed;
-  }
-  
-  /**
-   * Execute.
-   */
-  function execute($node) {
-    if (!isset($node->machine_name)) {
-      return;
-    }
-    if ($this->condition_used()) {
-      foreach ($this->get_contexts() as $context) {
-        $machine_names = $this->fetch_from_context($context, 'values');
-
-        if (in_array($node->machine_name, $machine_names)) {
-          $this->condition_met($context);
-        }
-      }
-    }
-  }
-}
\ No newline at end of file
