diff --git a/file_entity.pages.inc b/file_entity.pages.inc
index e9ae2b4..e8006f7 100644
--- a/file_entity.pages.inc
+++ b/file_entity.pages.inc
@@ -852,7 +852,7 @@ function file_entity_edit_submit($form, &$form_state) {
   if (!empty($form_state['values']['replace_upload'])) {
     $replacement = $form_state['values']['replace_upload'];
     // Move file from temp to permanent home.
-    $destination_uri = drupal_dirname($file->uri) . drupal_basename($replacement->uri);
+    $destination_uri = rtrim($file->uri, drupal_basename($file->uri)) . drupal_basename($replacement->uri);
     $replace_mode = $destination_uri == $file->uri ? FILE_EXISTS_REPLACE : FILE_EXISTS_RENAME;
     if ($new_file_uri = file_unmanaged_copy($replacement->uri, $destination_uri, $replace_mode)) {
       // @todo Add watchdog() about replaced file here?
