diff --git a/MediaYouTubeStreamWrapper.inc b/MediaYouTubeStreamWrapper.inc
index d6833e9..76d6f22 100644
--- a/MediaYouTubeStreamWrapper.inc
+++ b/MediaYouTubeStreamWrapper.inc
@@ -5,4 +5,4 @@
  *   updated to 1.0-alpha5 or later. This is retained until then, so that
  *   update.php can bootstrap.
  */
-include_once 'includes/MediaYouTubeStreamWrapper.inc';
+include_once DRUPAL_ROOT . '/' . drupal_get_path('module', 'media_youtube') . '/includes/includes/MediaYouTubeStreamWrapper.inc';
diff --git a/media_youtube.info b/media_youtube.info
index 1d766a3..d98705e 100644
--- a/media_youtube.info
+++ b/media_youtube.info
@@ -4,8 +4,10 @@ description = Provides YouTube support to the Media module.
 package = "Media"
 core = 7.x
 files[] = media_youtube.module
-files[] = includes/MediaYouTubeStreamWrapper.inc
 files[] = includes/MediaInternetYouTubeHandler.inc
-files[] = includes/media_youtube.styles.inc
+files[] = includes/MediaYouTubeStreamWrapper.inc
 files[] = includes/MediaYouTubeStyles.inc
+files[] = includes/media_youtube.formatters.inc
+files[] = includes/media_youtube.styles.inc
+files[] = includes/media_youtube.variables.inc
 dependencies[] = media_internet
diff --git a/media_youtube.install b/media_youtube.install
index 48d23ed..768038a 100644
--- a/media_youtube.install
+++ b/media_youtube.install
@@ -16,6 +16,8 @@ function media_youtube_install() {
  * Implement hook_uninstall().
  */
 function media_youtube_uninstall() {
+  drupal_load('module', 'media_youtube');
+
   foreach (media_youtube_variable_default() as $variable => $value) {
     media_youtube_variable_del($variable);
   }
diff --git a/media_youtube.module b/media_youtube.module
index 2282169..071812c 100644
--- a/media_youtube.module
+++ b/media_youtube.module
@@ -15,16 +15,16 @@
  */
 
 // A registry of variable_get defaults.
-include_once('includes/media_youtube.variables.inc');
+include_once DRUPAL_ROOT . '/' . drupal_get_path('module', 'media_youtube') . '/includes/media_youtube.variables.inc';
 
 // Hooks and callbacks for integrating with Styles module for display.
 // @todo Can save a little overhead for people without Styles module by wrapping
 //   this inside a module_exists('styles'). However, is that safe to do in
 //   global context? If not, is there any down side to doing it in hook_init()?
-include_once('includes/media_youtube.styles.inc');
+include_once DRUPAL_ROOT . '/' . drupal_get_path('module', 'media_youtube') . '/includes/media_youtube.styles.inc';
 
 // Hooks and callbacks for integrating with File Entity module for display.
-include_once('includes/media_youtube.formatters.inc');
+include_once DRUPAL_ROOT . '/' . drupal_get_path('module', 'media_youtube') . '/includes/media_youtube.formatters.inc';
 
 /**
  * Implements hook_media_internet_providers().
