diff --git a/cas.module b/cas.module
index 79a0c9e..f1ced80 100644
--- a/cas.module
+++ b/cas.module
@@ -99,7 +99,7 @@ function cas_login_check($force_authentication = TRUE) {
   if (empty($cas_user['login']) || empty($cas_user['name'])) {
     // Only set a warning if we forced login.
     if ($force_authentication) {
-      drupal_set_message(t('The user account %name is not available on this site.', array('%name' => $cas_name)), 'error');
+      drupal_set_message(t('The user account %name is not available on this site.', array('%name' => $cas_user['name'])), 'error');
     }
     return;
   }
